pytorch_addons/docs/bladedisc_torch_overview.md

```diff
@@ -0,0 +1,112 @@
+MHLO was chosen to be the tensor level "hub" IR between PyTorch and BladeDISC
+compiler world, with the considerations:
+
+- Dynamic shape support
+- Control flow support
+- Available and well established BladeDISC compiler
+- Continuous improvements and updates from the MLIR/XLA community
+
+### 2. Conversion & Lowering
+
+During optimization and compilation, there are many conversions and lowering
+among TorchScript and MHLO operators. We had to do the following things:
+
+- Inline the functions and methods
+- Freeze the parameters while reserving the required attributes
+- Reduce and concrete the ranks and data types of the tensor and parameters
+- Do constant folding and remove irrelevance or dead codes
+- Do aliasing analysis and rewrite in-place operations to out-of-place
+  corresponding versions
+
+### 3. Converters
+
+The IR converters are exactly the main mechanism that supports converting and
+lowering from TorchScript to MHLO
+
+- Query if the conversion of a graph node would succeed under a certain
+  conversion context
+- Build an MHLO module during converting TorchScript subgraph to MHLO
+
+### 4. Clustering and fallback
+
+TorchScript language is a subset of python for Deep Learning. There are many
+primitive operators and data structures that are reflected by python. MHLO can
+only support a subset of operators of TorchScript where tensors are values that
+can't be modified. Those issues are called *language subset problems*. BladeDISC
+addresses such problems with a clustering and fallback mechanism:
+
+- Iteratively, clean pythonic IR's and reduce the TorchScript graph closer to
+  computations with SSA semantics
+- Collect supported candidate nodes to backend with the simplified graph
+- Do clustering on the supported subset of nodes; The cluster had to be canceled
+  if a cyclic was constructed
+- Finally, fuse the clustered subgraph and lower it to the compiler backend; All
+  the nodes outside the clusters are fallback to TorchScript.
+
+### 5. Runtime(RAL, Runtime Abstraction Layer)
+
+The runtime of our BladeDISC compiler, [more details](<>)
+
+## Some additional views
+
+### 1. Relations between aten::Tensor and mlir::Value(Tensor)
+
+An `aten::Tensor` actually is a memory buffer that are shaped, so that it could
+be viewed, modified in place and slices. However, `mlir::Value` could not be
+modified. Any operations on `mlir::Value` would create new `mlir::Value`. The
+memory buffer was neglected until graph was lowering to LMHLO.
+
+### 2. Relations between Scalar and 0-rank Tensor
+
+- A torch aten tensor can be viewed, reshaped, sliced and modified. But a torch
+  Scalar can't.
+- A 0-rank tensor in MHLO is equivalent to a scalar, and they could be converted
+  to each other
+- In BladeDISC, torch Scalar was converted to mlir standard dialect operator
+  mlir::ConstantOp
+
+### 3. How BladeDISC converts List\[Scalar\] and List\[Tensor\]
+
+BladeDISC only converts those Lists of Scalars and Tensors that whose elements
```

Comment:
BladeDISC only converts ~those~ Lists of Scalars and Tensors ~that~ whose elements